README.md
# Flask-WhooshAlchemy3
[![PyPI version](https://badge.fury.io/py/flask-whooshalchemy3.svg)](https://badge.fury.io/py/flask-whooshalchemy3)
[![license](https://img.shields.io/github/license/blakev/flask-whooshalchemy3.svg)]()

Whoosh indexing capabilities for Flask-SQLAlchemy, Python 3 compatibility fork.
Performance improvements and suggestions are readily welcome.

Inspired from gyllstromk's [Flask-WhooshAlchemy](https://github.com/gyllstromk/Flask-WhooshAlchemy).

- [Whoosh](http://whoosh.readthedocs.io/en/latest/intro.html)
- [Flask-SqlAlchemy](http://flask-sqlalchemy.pocoo.org/2.1/)


## Install

```bash
$ pip install flask-whooshalchemy3
```

..alternatively from source,

```bash
$ pip install git+git://github.com/blakev/Flask-WhooshAlchemy3.git@master
```


## Quickstart

```python

from datetime import datetime

import flask_sqlalchemy
import flask_whooshalchemy
from whoosh.analysis import StemmingAnalyzer

db = flask_sqlalchemy.SQLAlchemy()

class BlogPost(db.Model):
    __tablename__ = 'posts'
    __searchable__ = ['title', 'content', 'summary']  # indexed fields
    __analyzer__ = StemmingAnalyzer()
    
    id = db.Column(db.Integer, primary_key=True)
    title = db.Column(db.String(255), unique=True)
    content = db.Column(db.Text(32*1024))
    summary = db.Column(db.String(1024))
    created = db.Column(db.DateTime, default=datetime.utcnow)

```

Committing model instances to the session will write or update the Whoosh index.

```python
db.session.add(BlogPost(title='First Post!', content='This is awesome.'))
db.session.commit()
```

Searching is done via `Model.query.search(..)`. However, the request must be done within the Flask
request context otherwise the database connection may not be established.

```python
@app.route('/posts')
def posts():
    num_posts = min(request.args.get('limit', 10), 50)
    query = request.args.get('q', '')
    results = BlogPost.query.search(query, limit=num_posts)
```


Results are ordered by Whoosh's ranking-algorithm, but can be overwritten with SQLAlchemy `.order_by`.

```python
yesterday = datetime.utcnow() - timedelta(days=1)
results = BlogPost.query
            .filter(BlogPost.created > yesterday)
            .search('first')
            .order_by(desc(BlogPost.created))
```

## Flask Configuration

`WHOOSH_ANALYZER` **(whoosh.Analyzer)**
- Sets the global text analyzer, available options [in Whoosh documentation](http://whoosh.readthedocs.io/en/latest/analysis.html). 
- Default: `StemmingAnalyzer`.

`WHOOSH_INDEX_PATH` (str)
- File path to where the text indexes will be saved. 
- Default: `{cwd}/.indexes/*`

`WHOOSH_INDEXING_CPUS` (int)
- The number of system processes to spawn for indexing new and modified documents.
- Default: `2`

`WHOOSH_INDEXING_RAM` (int)
- The amount of RAM, in megabytes, to reserve per indexing process for data processing.
- Default: `256`

`WHOOSH_RAM_CACHE` (bool)
- Allows common queries and their fields to be stored in cache, in RAM.
- Default: `False`

## License

```text
MIT License

Copyright (c) 2017 Blake VandeMerwe

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.
```

ChangeLog
CHANGES
=======

0.2.0
-----

* Set description format to markdown
* Fix #2 #3: Update version dependencies, support python3.8
* Fix #2: AttributeError 'BaseQuery' object has no 'search'
* Fix 1: BaseQuery object has no attribute search
* Update README

0.1.1
-----

* Misc changes
* Project restructure
* Update README.md
* Update requirements to include pbr

0.1.0
-----

* First!
* update .gitignore
* Create README.md
* Initial commit


tests/test_all.py
#! /usr/bin/env python
# -*- coding: utf-8 -*-
# >>
#     Flask-WhooshAlchemy3, 2017
# <<


from flask import Flask
from flask_sqlalchemy import SQLAlchemy
from flask_testing import TestCase
import flask_whooshalchemy as wa
from whoosh.analysis import StemmingAnalyzer, DoubleMetaphoneFilter

import datetime
import os
import tempfile
import shutil

db = SQLAlchemy()


class BlogishBlob(object):
    id = db.Column(db.Integer, primary_key=True)
    title = db.Column(db.Text)
    content = db.Column(db.String)
    blurb = db.Column(db.String)
    ignored = db.Column(db.String)
    created = db.Column(db.DateTime(), default=datetime.datetime.utcnow())

    def __repr__(self):
        return '{0}(title={1})'.format(self.__class__.__name__, self.title)


def _after_flush(app, changes):
    from sqlalchemy.orm import EXT_CONTINUE
    return EXT_CONTINUE


class ObjectA(db.Model, BlogishBlob):
    __tablename__ = 'objectA'
    __searchable__ = ['title', 'content', 'blurb']


class ObjectB(db.Model, BlogishBlob):
    __tablename__ = 'objectB'
    __searchable__ = ['title', 'content', 'content']  # dup intentional


class ObjectC(db.Model, BlogishBlob):
    __tablename__ = 'objectC'
    __searchable__ = ['title', 'field_that_doesnt_exist']


class ObjectD(db.Model, BlogishBlob):
    __tablename__ = 'objectD'
    __searchable__ = ['title']
    __analyzer__ = StemmingAnalyzer() | DoubleMetaphoneFilter()


class Tests(TestCase):
    DATABASE_URL = 'sqlite://'
    TESTING = True

    def create_app(self):
        tmp_dir = tempfile.mkdtemp()

        app = Flask(__name__)

        app.config['WHOOSH_BASE'] = os.path.join(tmp_dir, 'whoosh')

        return app

    def setUp(self):
        db.init_app(self.app)
        db.create_all()

    def tearDown(self):
        try:
            shutil.rmtree(self.app.config['WHOOSH_BASE'])
        except OSError as e:
            if e.errno != 2:  # code 2 - no such file or directory
                raise

        db.drop_all()

    def test_all(self):
        title1 = u'a slightly long title'
        title2 = u'another title'
        title3 = u'wow another title'

        x = ObjectA
        obj = ObjectA(title='title', blurb='this is a blurb')
        db.session.add(obj)
        db.session.commit()

        self.assertEqual(len(list(ObjectA.query.search('blurb'))), 1)
        db.session.delete(obj)
        db.session.commit()

        db.session.add(
            ObjectA(title=title1, content=u'hello world', ignored=u'no match'))
        db.session.commit()

        self.assertEqual(len(list(ObjectA.query.search('what'))), 0)
        self.assertEqual(len(list(ObjectA.query.search(u'no match'))),
                         0)
        self.assertEqual(len(list(ObjectA.query.search(u'title'))), 1)
        self.assertEqual(len(list(ObjectA.query.search(u'hello'))), 1)

        db.session.add(ObjectB(title=u'my title', content=u'hello world'))
        db.session.commit()

        # make sure does not interfere with ObjectA's results
        self.assertEqual(len(list(ObjectA.query.search(u'what'))), 0)
        self.assertEqual(len(list(ObjectA.query.search(u'title'))), 1)
        self.assertEqual(len(list(ObjectA.query.search(u'hello'))), 1)

        self.assertEqual(len(list(ObjectB.query.search(u'what'))), 0)
        self.assertEqual(len(list(ObjectB.query.search(u'title'))), 1)
        self.assertEqual(len(list(ObjectB.query.search(u'hello'))), 1)

        obj2 = ObjectA(title=title2, content=u'a different message')
        db.session.add(obj2)
        db.session.commit()

        self.assertEqual(len(list(ObjectA.query.search(u'what'))), 0)
        l = list(ObjectA.query.search(u'title'))
        self.assertEqual(len(l), 2)

        # ranking should always be as follows, since title2 should have a higher relevance score

        self.assertEqual(l[0].title, title2)
        self.assertEqual(l[1].title, title1)

        self.assertEqual(len(list(ObjectA.query.search(u'hello'))), 1)
        self.assertEqual(len(list(ObjectA.query.search(u'message'))), 1)

        self.assertEqual(len(list(ObjectB.query.search(u'what'))), 0)
        self.assertEqual(len(list(ObjectB.query.search(u'title'))), 1)
        self.assertEqual(len(list(ObjectB.query.search(u'hello'))), 1)
        self.assertEqual(len(list(ObjectB.query.search(u'message'))), 0)

        db.session.add(ObjectA(title=title3, content=u'a different message'))
        db.session.commit()

        l = list(ObjectA.query.search(u'title'))
        self.assertEqual(len(l), 3)
        self.assertEqual(l[0].title, title2)
        self.assertEqual(l[1].title, title1)
        self.assertEqual(l[2].title, title3)

        db.session.delete(obj2)
        db.session.commit()

        l = list(ObjectA.query.search(u'title'))
        self.assertEqual(len(l), 2)
        self.assertEqual(l[0].title, title1)
        self.assertEqual(l[1].title, title3)

        two_days_ago = datetime.date.today() - datetime.timedelta(2)

        title4 = u'a title that is significantly longer than the others'

        db.session.add(ObjectA(title=title4, created=two_days_ago))
        db.session.commit()

        one_day_ago = datetime.date.today() - datetime.timedelta(1)

        recent = list(ObjectA.query.search(u'title')
                      .filter(ObjectA.created >= one_day_ago))

        self.assertEqual(len(recent), 2)
        self.assertEqual(l[0].title, title1)
        self.assertEqual(l[1].title, title3)

        three_days_ago = datetime.date.today() - datetime.timedelta(3)

        l = list(ObjectA.query.search(u'title')
                 .filter(ObjectA.created >= three_days_ago))

        self.assertEqual(len(l), 3)
        self.assertEqual(l[0].title, title1)
        self.assertEqual(l[1].title, title3)
        self.assertEqual(l[2].title, title4)

        title5 = u'title with title as frequent title word'

        db.session.add(ObjectA(title=title5))
        db.session.commit()

        l = list(ObjectA.query.search(u'title'))
        self.assertEqual(len(l), 4)
        self.assertEqual(l[0].title, title5)
        self.assertEqual(l[1].title, title1)
        self.assertEqual(l[2].title, title3)
        self.assertEqual(l[3].title, title4)

        # test limit
        l = list(ObjectA.query.search(u'title', limit=2))
        self.assertEqual(len(l), 2)
        self.assertEqual(l[0].title, title5)
        self.assertEqual(l[1].title, title1)

        # XXX should replace this with a new function, but I can't figure out
        # how to do this cleanly with flask sqlalchemy and testing

        db.drop_all()
        db.create_all()

        title1 = u'my title'
        db.session.add(ObjectA(title=title1, content=u'hello world'))
        db.session.commit()

        l = list(ObjectA.query.search(u'title'))
        self.assertEqual(len(l), 1)

        l = list(ObjectA.query.search(u'hello'))
        self.assertEqual(len(l), 1)

        l = list(ObjectA.query.search(u'title', fields=('title',)))
        self.assertEqual(len(l), 1)
        l = list(ObjectA.query.search(u'hello', fields=('title',)))
        self.assertEqual(len(l), 0)

        l = list(ObjectA.query.search(u'title', fields=('content',)))
        self.assertEqual(len(l), 0)
        l = list(ObjectA.query.search(u'hello', fields=('content',)))
        self.assertEqual(len(l), 1)

        l = list(
            ObjectA.query.search(u'hello dude', fields=('content',),
                                        or_=True))
        self.assertEqual(len(l), 1)

        l = list(
            ObjectA.query.search(u'hello dude', fields=('content',),
                                        or_=False))
        self.assertEqual(len(l), 0)

        # new function: test chaining
        db.drop_all()
        db.create_all()

        db.session.add(ObjectA(title=u'title one', content=u'a poem'))
        db.session.add(ObjectA(title=u'title two', content=u'about testing'))
        db.session.add(
            ObjectA(title=u'title three', content=u'is delightfully tested'))
        db.session.add(ObjectA(title=u'four', content=u'tests'))
        db.session.commit()

        self.assertEqual(len(list(ObjectA.query.search(u'title'))), 3)
        self.assertEqual(len(list(ObjectA.query.search(u'test'))), 3)

        # chained query, operates as AND
        self.assertEqual(len(list(
            ObjectA.query.search(u'title').search(u'test'))),
                         2)

    def test_invalid_attribute(self):
        db.session.add(ObjectC(title=u'my title', content=u'hello world'))
        self.assertRaises(AttributeError, db.session.commit)

    def test_default_analyzer(self):
        db.session.add(ObjectA(title=u'jumping', content=u''))
        db.session.commit()
        assert ['jumping'] == [obj.title for obj in
                               ObjectA.query.search(u'jump')]

    def test_custom_analyzer(self):
        from whoosh.analysis import SimpleAnalyzer
        self.app.config['WHOOSH_ANALYZER'] = SimpleAnalyzer()
        db.init_app(self.app)
        db.create_all()
        db.session.add(ObjectA(title=u'jumping', content=u''))
        db.session.commit()
        assert not list(ObjectA.query.search(u'jump'))
        assert ['jumping'] == [obj.title for obj in
                               ObjectA.query.search(u'jumping')]

        db.session.add(ObjectD(title=u'Travelling', content=u'Stemming'))
        db.session.add(
            ObjectD(title=u'travel', content=u'Unstemmed and normal'))
        db.session.add(ObjectD(title=u'trevel', content=u'Mispelt'))

        db.session.commit()
        # When mispelt on either the indexed side or the query side, they should all return 3 due to the DoubleMetaphoneFilter
        self.assertEqual(len(list(ObjectD.query.search('travelling'))),
                         3)
        self.assertEqual(len(list(ObjectD.query.search('trovel'))), 3)


if __name__ == '__main__':
    import unittest
    unittest.main()


scratch.py
#! /usr/bin/env python
# -*- coding: utf-8 -*-
# >>
#     Flask-WhooshAlchemy3, 2017
# <<

import random
from flask import Flask, request
from flask_sqlalchemy import SQLAlchemy
import flask_whooshalchemy

from whoosh.analysis import StemmingAnalyzer
app = Flask(__name__)
app.config.update(dict(
    SQLALCHEMY_DATABASE_URI='sqlite:///db.sqlite',
    SQL_TRACK_MODIFICATIONS=True,
    WHOOSH_INDEX_PATH='whooshIndex',
    WHOOSH_ANALYZER='StemmingAnalyzer',
))

db = SQLAlchemy()


class Org(db.Model):
    __searchable__ = ['orgName', 'phone1', 'phone2', 'phone3', 'email1', 'email2', 'email3', 'backgroundInfo']
    __analyzer__ = StemmingAnalyzer()
    __tablename__ = 'Org'

    id = db.Column(db.Integer, primary_key=True)
    active = db.Column(db.Boolean, unique=False, default=True)
    orgName = db.Column(db.String(200), unique=True, nullable=False)
    postalStreet = db.Column(db.String(200))
    postalTown = db.Column(db.String(60))
    postalState = db.Column(db.String(20))
    postalPostCode = db.Column(db.String(10))
    backgroundInfo = db.Column(db.String(1000))
    phone1 = db.Column(db.String(128))
    phone2 = db.Column(db.String(128))
    phone3 = db.Column(db.String(128))
    phone1Type = db.Column(db.String(24))
    phone2Type = db.Column(db.String(24))
    phone3Type = db.Column(db.String(24))
    email1 = db.Column(db.String(128))
    email2 = db.Column(db.String(128))
    email3 = db.Column(db.String(128))
    email1Type = db.Column(db.String(24))
    email2Type = db.Column(db.String(24))
    email3Type = db.Column(db.String(24))

    def __repr__(self):  # this is how our object is printed when we print it out
        return f"Org('{self.orgName}')"


@app.route('/search')
def search():
    num_posts = min(int(request.args.get('limit', 10)), 50)
    query = request.args.get('q', '')
    results = Org.query.search(query, limit=num_posts)
    return '\n'.join(map(str, results))


BOOL = (True, False,)
STATES = ('UT', 'ID', 'CA', 'CO', 'WY', 'AR', 'NV',)
NAMES = ('Red', 'Green', 'Blue', 'Yellow',)

@app.before_first_request
def bootstrap():
    db.create_all()
    flask_whooshalchemy.search_index(app, Org)

    for x in range(100):
        rec = Org(active=random.choice(BOOL),
                  orgName=f'{random.choice(NAMES)}_{x}',
                  postalState=random.choice(STATES),
                  postalStreet=random.randrange(100, 1500))
        db.session.add(rec)
    db.session.commit()


if __name__ == "__main__":
    db.init_app(app)
    app.run()




flask_whooshalchemy.py
#! /usr/bin/env python
# -*- coding: utf-8 -*-
# >>
#     Flask-WhooshAlchemy3, 2017
# <<

import os
import heapq
import logging
from collections import defaultdict

import whoosh
import sqlalchemy
import flask_sqlalchemy
import whoosh.index
from whoosh import fields as whoosh_fields
from whoosh.analysis import StemmingAnalyzer
from whoosh.qparser import OrGroup, AndGroup, MultifieldParser
from whoosh.filedb.filestore import RamStorage
from whoosh.writing import AsyncWriter
from sqlalchemy import types as sql_types
from sqlalchemy.orm import EXT_CONTINUE

logger = logging.getLogger(__name__)


# DEFAULTS
DEFAULT_WHOOSH_ANALYZER = StemmingAnalyzer()
DEFAULT_WHOOSH_INDEX_PATH = os.path.join(
    os.path.abspath(os.getcwd()), '.indexes')

UPDATE_FIELDS = ('update', 'insert')
TEXT_TYPES = (sql_types.String, sql_types.Unicode, sql_types.Text)
DATE_TYPES = (sql_types.DateTime, sql_types.Date)
NUM_TYPES = (sql_types.Integer, sql_types.BigInteger, sql_types.SmallInteger,
             sql_types.Float, sql_types.Binary)


class WhooshAlchemyError(Exception):
    """ Base exception class for Flask-WhooshAlchemy3 """


class QueryProxy(flask_sqlalchemy.BaseQuery):
    def __init__(self, entities, session=None):
        super(QueryProxy, self).__init__(entities, session)

        # sqlalchemy database model base class
        self._model_cls = self._mapper_zero().class_

        # references into our Whoosh index
        self._pk = self._model_cls.whoosh_pk
        self._searcher = self._model_cls.whoosh

        # place to store our whoosh results against db results
        self._whoosh_results = None

    def __iter__(self):
        """ Sort the results by Whoosh rank; relevance. """
        _iter = super(QueryProxy, self).__iter__()

        if self._whoosh_results is None or self._order_by is not False:
            return _iter

        ordered = []

        for row in _iter:
            # we have to convert the primary-key, as stored in the SQL database
            #  into a string because the key is stored as an `ID` in whoosh.
            #  The ID field is string only; plus, this allows for uuid pk's.
            str_pk = str(getattr(row, self._pk))
            heapq.heappush(
                ordered, (self._whoosh_results[str_pk], row))

        def inner():
            while ordered:
                yield heapq.heappop(ordered)[1]
        return inner()

    def search(self, query, limit=None, fields=None, or_=False):
        """ Perform a woosh index search. """

        if not isinstance(query, str):
            raise WhooshAlchemyError('query parameter must be string-like')

        results = self._searcher(query, limit, fields, or_)

        if not results:
            return self.filter(sqlalchemy.text('null'))

        result_set = set()
        result_rank = {}

        for rank, result in enumerate(results):
            result_set.add(result)
            result_rank[result] = rank

        f = self.filter(getattr(self._model_cls, self._pk).in_(result_set))
        f._whoosh_results = result_rank
        return f


class Searcher(object):
    __slots__ = ('pk', 'index', 'searcher', 'fields')

    def __init__(self, primary_key, index):
        self.pk = primary_key
        self.index = index
        self.searcher = self.index.searcher
        self.fields = list(set(index.schema._fields.keys()) - {self.pk})

    def __call__(self, query, limit=None, fields=None, or_=False):
        if fields is None:
            fields = self.fields
        group = OrGroup if or_ else AndGroup
        parser = MultifieldParser(fields, self.index.schema, group=group)
        results = []
        with self.searcher() as searcher:
            for doc in searcher.search(parser.parse(query), limit=limit):
                results.append(doc[self.pk])
        return results


def _post_flush(app, changes):
    by_type = defaultdict(list)

    for instance, change in changes:
        update = change in UPDATE_FIELDS

        if hasattr(instance.__class__, '__searchable__'):
            by_type[instance.__class__].append((update, instance))

    procs = app.config.get('WHOOSH_INDEXING_CPUS', 2)
    limit = app.config.get('WHOOSH_INDEXING_RAM', 256)

    for model_name, values in by_type.items():
        ref = values[0][1]

        index = AsyncWriter(search_index(app, ref),
                            delay=0.15,
                            writerargs=dict(proc=procs, limitmb=limit))
        primary_field = ref.whoosh.pk
        searchable = ref.__searchable__

        with index as writer:
            for do_update, v in values:
                pk = str(getattr(v, primary_field))

                if do_update:
                    # get all the columns defined in `__searchable__`
                    attrs = {}
                    for k in searchable:
                        try:
                            attrs[k] = str(getattr(v, k))
                        except AttributeError:
                            raise AttributeError('invalid attribute `%s`' % k)

                    attrs[primary_field] = pk

                    # create a new document, or update an old one, with all
                    #  of our new column values
                    writer.update_document(**attrs)
                else:
                    # remove the document by field `primary field` value `pk`
                    writer.delete_by_term(primary_field, pk)
    return EXT_CONTINUE


def create_index(app, model):
    path = app.config.get('WHOOSH_INDEX_PATH', DEFAULT_WHOOSH_INDEX_PATH)
    name = model.__tablename__
    # this is where all the search indexes will be stored on disk
    full_path = os.path.join(path, name)

    analyzer = get_analyzer(app, model)
    schema, pk = get_schema(model, analyzer)

    if whoosh.index.exists_in(full_path):
        index = whoosh.index.open_dir(full_path, schema=schema)

    else:
        if not os.path.exists(full_path):
            os.makedirs(full_path)
        index = whoosh.index.create_in(full_path, schema)

    app.search_indexes[name] = index

    kls = model.__class__
    kls.whoosh = Searcher(pk, index)
    kls.whoosh_pk = pk
    kls.query_class = QueryProxy

    if app.config.get('WHOOSH_RAM_CACHE', False):
        kls.whoosh.searcher.set_caching_policy(storage=RamStorage())
    return index


def get_analyzer(app, model):
    analyzer = getattr(model, '__analyzer__', None)
    if not analyzer:
        analyzer = app.config.get('WHOOSH_ANALYZER', DEFAULT_WHOOSH_ANALYZER)
        setattr(model, '__analyzer__', analyzer)
    return analyzer


def get_schema(model, analyzer):
    schema = {}
    primary = None
    searchable = set(getattr(model, '__searchable__', []))

    for field in model.__table__.columns:
        # primary key id
        if field.primary_key:
            schema[field.name] = whoosh_fields.ID(
                stored=True, unique=True, sortable=True)
            primary = field.name

        if field.name not in searchable:
            continue

        # text types
        if isinstance(field.type, TEXT_TYPES):
            schema[field.name] = whoosh_fields.TEXT(analyzer=analyzer)

        elif isinstance(field.type, DATE_TYPES):
            is_unique = getattr(field, 'unique', False)
            schema[field.name] = whoosh_fields.DATETIME(unique=is_unique)

        elif isinstance(field.type, sql_types.Boolean):
            schema[field.name] = whoosh_fields.BOOLEAN()

        else:
            raise WhooshAlchemyError(
                'cannot index column of type %s' % field.type)

    return whoosh_fields.Schema(**schema), primary


def search_index(app, model):
    if not hasattr(app, 'search_indexes'):
        app.search_indexes = {}
    index = app.search_indexes.get(
        model.__tablename__, create_index(app, model))
    return index

flask_sqlalchemy.models_committed.connect(_post_flush)


flask_whooshalchemy/__init__.py
#! /usr/bin/env python
# -*- coding: utf-8 -*-
# >>
#     Flask-WhooshAlchemy3, 2017
# <<

import os
import heapq
import logging
from collections import defaultdict

import whoosh
import sqlalchemy
import flask_sqlalchemy
import whoosh.index
from whoosh import fields as whoosh_fields
from whoosh.analysis import StemmingAnalyzer
from whoosh.qparser import OrGroup, AndGroup, MultifieldParser
from whoosh.filedb.filestore import RamStorage
from whoosh.writing import AsyncWriter
from sqlalchemy import types as sql_types
from sqlalchemy.orm import EXT_CONTINUE

logger = logging.getLogger(__name__)


# DEFAULTS
DEFAULT_WHOOSH_ANALYZER = StemmingAnalyzer()
DEFAULT_WHOOSH_INDEX_PATH = os.path.join(
    os.path.abspath(os.getcwd()), '.indexes')

UPDATE_FIELDS = ('update', 'insert')
TEXT_TYPES = (sql_types.String, sql_types.Unicode, sql_types.Text)
DATE_TYPES = (sql_types.DateTime, sql_types.Date)
NUM_TYPES = (sql_types.Integer, sql_types.BigInteger, sql_types.SmallInteger,
             sql_types.Float, sql_types.Binary)


class WhooshAlchemyError(Exception):
    """ Base exception class for Flask-WhooshAlchemy3 """


class QueryProxy(flask_sqlalchemy.BaseQuery):
    def __init__(self, entities, session=None):
        super(QueryProxy, self).__init__(entities, session)

        # sqlalchemy database model base class
        self._model_cls = self._mapper_zero().class_

        # references into our Whoosh index
        self._pk = self._model_cls.whoosh_pk
        self._searcher = self._model_cls.whoosh

        # place to store our whoosh results against db results
        self._whoosh_results = None

    def __iter__(self):
        """ Sort the results by Whoosh rank; relevance. """
        _iter = super(QueryProxy, self).__iter__()

        if self._whoosh_results is None or self._order_by is not False:
            return _iter

        ordered = []

        for row in _iter:
            # we have to convert the primary-key, as stored in the SQL database
            #  into a string because the key is stored as an `ID` in whoosh.
            #  The ID field is string only; plus, this allows for uuid pk's.
            str_pk = str(getattr(row, self._pk))
            heapq.heappush(
                ordered, (self._whoosh_results[str_pk], row))

        def inner():
            while ordered:
                yield heapq.heappop(ordered)[1]
        return inner()

    def search(self, query, limit=None, fields=None, or_=False):
        """ Perform a woosh index search. """

        if not isinstance(query, str):
            raise WhooshAlchemyError('query parameter must be string-like')

        results = self._searcher(query, limit, fields, or_)

        if not results:
            return self.filter(sqlalchemy.text('null'))

        result_set = set()
        result_rank = {}

        for rank, result in enumerate(results):
            result_set.add(result)
            result_rank[result] = rank

        f = self.filter(getattr(self._model_cls, self._pk).in_(result_set))
        f._whoosh_results = result_rank
        return f


class Searcher(object):
    __slots__ = ('pk', 'index', 'searcher', 'fields')

    def __init__(self, primary_key, index):
        self.pk = primary_key
        self.index = index
        self.searcher = self.index.searcher
        self.fields = list(set(index.schema._fields.keys()) - {self.pk})

    def __call__(self, query, limit=None, fields=None, or_=False):
        if fields is None:
            fields = self.fields
        group = OrGroup if or_ else AndGroup
        parser = MultifieldParser(fields, self.index.schema, group=group)
        results = []
        with self.searcher() as searcher:
            for doc in searcher.search(parser.parse(query), limit=limit):
                results.append(doc[self.pk])
        return results


def _post_flush(app, changes):
    by_type = defaultdict(list)

    for instance, change in changes:
        update = change in UPDATE_FIELDS

        if hasattr(instance.__class__, '__searchable__'):
            by_type[instance.__class__].append((update, instance))

    procs = app.config.get('WHOOSH_INDEXING_CPUS', 2)
    limit = app.config.get('WHOOSH_INDEXING_RAM', 256)

    for model_name, values in by_type.items():
        ref = values[0][1]

        index = AsyncWriter(search_index(app, ref),
                            delay=0.15,
                            writerargs=dict(proc=procs, limitmb=limit))
        primary_field = ref.whoosh.pk
        searchable = ref.__searchable__

        with index as writer:
            for do_update, v in values:
                pk = str(getattr(v, primary_field))

                if do_update:
                    # get all the columns defined in `__searchable__`
                    attrs = {}
                    for k in searchable:
                        try:
                            attrs[k] = str(getattr(v, k))
                        except AttributeError:
                            raise AttributeError('invalid attribute `%s`' % k)

                    attrs[primary_field] = pk

                    # create a new document, or update an old one, with all
                    #  of our new column values
                    writer.update_document(**attrs)
                else:
                    # remove the document by field `primary field` value `pk`
                    writer.delete_by_term(primary_field, pk)
    return EXT_CONTINUE


def create_index(app, model):
    path = app.config.get('WHOOSH_INDEX_PATH', DEFAULT_WHOOSH_INDEX_PATH)
    name = model.__tablename__
    # this is where all the search indexes will be stored on disk
    full_path = os.path.join(path, name)

    analyzer = get_analyzer(app, model)
    schema, pk = get_schema(model, analyzer)

    if whoosh.index.exists_in(full_path):
        index = whoosh.index.open_dir(full_path, schema=schema)

    else:
        if not os.path.exists(full_path):
            os.makedirs(full_path)
        index = whoosh.index.create_in(full_path, schema)

    app.search_indexes[name] = index
    kls = model.__class__

    kls.whoosh = model.whoosh = Searcher(pk, index)
    kls.whoosh_pk = model.whoosh_pk = pk
    kls.query_class = model.query_class = QueryProxy

    if app.config.get('WHOOSH_RAM_CACHE', False):
        ram_storage = RamStorage()
        model.whoosh.searcher.set_caching_policy(storage=ram_storage)
        kls.whoosh.searcher.set_caching_policy(storage=ram_storage)
    return index


def get_analyzer(app, model):
    analyzer = getattr(model, '__analyzer__', None)
    if not analyzer:
        analyzer = app.config.get('WHOOSH_ANALYZER', DEFAULT_WHOOSH_ANALYZER)
        setattr(model, '__analyzer__', analyzer)
    return analyzer


def get_schema(model, analyzer):
    schema = {}
    primary = None
    searchable = set(getattr(model, '__searchable__', []))

    for field in model.__table__.columns:
        # primary key id
        if field.primary_key:
            schema[field.name] = whoosh_fields.ID(
                stored=True, unique=True, sortable=True)
            primary = field.name

        if field.name not in searchable:
            continue

        # text types
        if isinstance(field.type, TEXT_TYPES):
            schema[field.name] = whoosh_fields.TEXT(analyzer=analyzer)

        elif isinstance(field.type, DATE_TYPES):
            is_unique = getattr(field, 'unique', False)
            schema[field.name] = whoosh_fields.DATETIME(unique=is_unique)

        elif isinstance(field.type, sql_types.Boolean):
            schema[field.name] = whoosh_fields.BOOLEAN()

        else:
            raise WhooshAlchemyError(
                'cannot index column of type %s' % field.type)

    return whoosh_fields.Schema(**schema), primary


def search_index(app, model):
    if not hasattr(app, 'search_indexes'):
        app.search_indexes = {}
    index = app.search_indexes.get(
        model.__tablename__, create_index(app, model))
    return index


flask_sqlalchemy.models_committed.connect(_post_flush)
